src/librustc/ty/mod.rs

```diff
@@ -2623,3 +2575,47 @@ impl<'a, 'gcx, 'tcx> TyCtxt<'a, 'gcx, 'tcx> {
         }
     }
 }
+
+fn associated_item<'a, 'tcx>(tcx: TyCtxt<'a, 'tcx, 'tcx>, def_id: DefId)
+    -> AssociatedItem
+{
+    let id = tcx.hir.as_local_node_id(def_id).unwrap();
+    let parent_id = tcx.hir.get_parent(id);
+    let parent_def_id = tcx.hir.local_def_id(parent_id);
+    let parent_item = tcx.hir.expect_item(parent_id);
+    match parent_item.node {
+        hir::ItemImpl(.., ref impl_trait_ref, _, ref impl_item_refs) => {
+            for impl_item_ref in impl_item_refs {
+                let assoc_item =
+                    tcx.associated_item_from_impl_item_ref(parent_def_id,
```

Comment:
This looks basically right. I would probably rewrite this `for` loop (and the one below) to something like:

```rust
if let Some(impl_item_ref) = impl_item_refs.iter().find(|i| i.id.node_id == id) {
    return tcx.associated_item_from_impl_item_ref(...); // same parameters you have already
}
```